src/armature/Generator.ts

```diff
@@ -1,38 +1,189 @@
     at: Point;
 };
 
+/**
+ * A cost function returns the cost of an entire model. The new nodes that were added
+ * since the last iteration of generation are passed to the cost function so that, if you can,
+ * you can just compute the incremental cost and add it to the previous cost, `instance.getCost()`.
+ */
+export type CostFn = (instance: GeneratorInstance, nodes: Node[]) => number;
+
+/**
+ * An instance of a generated model, possibly in the middle of generation.
+ */
+export class GeneratorInstance {
+    private model: Model = new Model();
+    private generator: Generator;
+    private costFn: CostFn;
+    private cost: number = 0;
+    private spawnPoints: SpawnPoint[] = [];
+    private random: RandomGenerator = Math.random;
+
+    /**
+     * Creates an instance of a generator, with a function specifying how to evaluate how "good"
+     * it is.
+     *
+     * @param {Generator} generator The generator this is an instance of.
+     * @param {CostFn} costFn A function to evaluate this instance.
+     */
+    constructor(generator: Generator, costFn: CostFn) {
+        this.generator = generator;
+        this.costFn = costFn;
+    }
+
+    /**
+     * A wrapper for `mode.add`, for convenience when defining rules.
+     *
+     * @param {Node} node The node to add to the model.
+     * @returns {Node} The node that was added.
+     */
+    public add(node: Node): Node {
+        return this.model.add(node);
+    }
+
+    /**
+     * @returns {GeneratorInstance} A flat copy of this generator instance.
+     */
+    public clone(): GeneratorInstance {
+        const cloned = new GeneratorInstance(this.generator, this.costFn);
+        cloned.model = this.model.clone();
+        cloned.cost = this.cost;
+        cloned.spawnPoints = [...this.spawnPoints];
+
+        return cloned;
+    }
+
+    /**
+     * @returns {Model} The model this instance has generated so far.
+     */
+    public getModel(): Model {
+        return this.model;
+    }
+
+    /**
+     * @returns {number} The current cost of the model the instance has generated so far.
+     */
+    public getCost(): number {
+        return this.cost;
+    }
+
+    /**
+     * @returns {number} How many spawn points the generator can choose from when growing the
+     * next generation.
+     */
+    public activeSpawnPoints(): number {
+        return this.spawnPoints.length;
+    }
+
+    /**
+     * Tells the generator that more components can be generated somewhere.
+     *
+     * @param {SpawnPoint} spawnPoint The name of the component to spawn and the point at which to
+     * spawn it.
+     */
+    public addDetail(spawnPoint: SpawnPoint) {
+        this.spawnPoints.push(spawnPoint);
+    }
+
+    /**
+     * Grows this instance, if possible, until a new shape is added.
+     */
+    public advance() {
+        const originalLength = this.model.nodes.length;
+
+        while (this.model.nodes.length === originalLength && this.spawnPoints.length > 0) {
```

Comment:
Law of Demeter makes me think 🤔 when seeing `this.model.nodes.length`. Will `nodes` always be defined on the `Model` class? Is there a case where it could be `null` or `undefined`?

Reply:
No cases where it'll be null/undefined (the types also help with that.) I kind of think of `Model` mostly as a plain old data class, like its purpose is just to be a flat array of nodes that you can copy quickly and add to. I can add a `.numNodes()` method if you think that's best though

Reply:
nawh, it's chill - this is gucci. I'm just being dense

Reply:
naw dw, it needed more explanation in order to be able to make a decision about it